### DISCUSSION TITLE
Implicit block scoped catch

### DISCUSSION THREAD

**Discussion by gulshan** [dotnet/csharplang#4543] [Language Ideas]
The idea roughly is to transform this-
```csharp
void MyFunc()
{
    OperationA();
    OperationB();
    catch(MyException1 me1)
    {
        Log(me1);
    }
    catch(MyException2 me2)
    {
        Log(me2);
    }
}
```
to this-
```csharp
void MyFunc()
{
    try
    {
        OperationA();
        OperationB();
    }
    catch(MyException1 me1)
    {
        Log(me1);
    }
    catch(MyException2 me2)
    {
        Log(me2);
    }
}
```
So, the idea is, there can be catch blocks without try blocks. These try-less catch blocks will then catch all the exceptions thrown from current enclosing block- which can be a method or a conditional block or a loop.

What's the benefit of getting rid of the `try` block?
- Coupling try blocks with catch blocks implicitly encourages the programmer to catch each exception-throwing statements separately and putting exception-handling code closer to the potential thrower. The nested try-catch blocks come from this mindset. This decoupled catch blocks may encourage to put exception-handling at the end of a block, which is most probably also end of some workflow. Also this may give put the idea in the novice programmers mind that, exception-handling should be done later and together, not sooner and separately.
- Another bonus will be, the main operational flow will have one less indentation.

I did not think about it that seriously though. Just sharing what came to my mind.

**Comment by gulshan** [comment]
Also, can we have the expression bodied catch constructs already in C#?
```csharp
void MyFunc()
{
    OperationA();
    OperationB();

    catch(MyException1 me1) => Log(me1);
    catch(MyException2 me2) => Log(me2);
}
```

**Reply by OdinAnDieFreude** [reply]
There's basically no reason for a catch to require a code block after it, C# could allow a statement just as easily.

Then, the catch part would be

    catch (MyException1 me1) Log(me1);
    catch (MyException2 me2) Log(me2);

Expression bodied catch would only be usefull, if you wanted the try-catch-block to return something, like:

    void MyFunc()
    {
      var result = try
      {
        OperationA();
        OperationB();
      }
      catch (MyException1 me1) => Log(me1);
      catch (MyException2 me2) => Log(me2);
    }

But I don't think that this would be of much use.

**Comment by theunrepentantgeek** [comment]
Best practice for **try**...**catch** blocks is to put only the necessary code within the **try** - if you're expecting an `IOException` then pretty much only the code doing I/O should be within the try. 

Eliminating the **try** part of the expression just makes code needlessly obfuscated.

**Reply by gulshan** [reply]
Exception means invalid workflow state. I can think of 3 types of exception handling (excluding the programming errors, where the code should be modified)-
- **Continue with default value:** Put some default values in the state to repair it and continue the workflow.
- **Stop:** Gracefully stop the current workflow (workflow here may be a few lines or may even be the full program/application). Log and show appropriate error messages.
- **Retry:** Try to recover the state by retrying the steps which caused the exception. And eventually stop(previous case) if unsuccessful for certain times.

For the "continue with default value" case of exception handling, the catch block should be very close to the potential exception throwing code and I agree the scope of the try block should be as narrow as possible for this case.

For the "stop" case, the exception should be handled at workflow root, which is usually a function. No need to minimize the scope of the try block here.

For the "retry" case, the retrying is most probably happening within a loop. Here the scope of the try block should be all of the loop. Limiting the scope may not be beneficial.

Proposed implicit block-scoped catches will be helpful in the last to cases, IMHO. It will reduce clutter and encourage to treat a natural code-block as the scope of exception handling, not explicitly creating the try scopes.

**Reply by gulshan** [reply]
After a little bit of searching, found two quotes regarding try-catch best practices, which I think aligns well with this proposal.

Robert C. Martin in 'Clean Code'-
>If the keyword 'try' exists in a function, it should be the very first word in the function and that there should be nothing after the catch/finally blocks.

Herb Sutter [here](http://www.gotw.ca/gotw/065.htm)-
>Actually, it turns out that exception safety is rarely about writing 'try' and 'catch' -- and the more rarely the better.

**Reply by HaloFour** [reply]
Those quotes don't apply to C#, nor should they.

**Reply by theunrepentantgeek** [reply]
If you're following Clean Code guidance, your methods will only be four or five statements at most. 

In such a situation the only code within the **try**..**catch** will be the operation throwing the exception, so my point still applies.
